packages/zen-bull-vault/src/EmergencyWithdraw.sol

```diff
@@ -0,0 +1,262 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity =0.7.6;
+pragma abicoder v2;
+
+import { console } from "forge-std/console.sol";
+
+// interface
+import { IERC20 } from "openzeppelin/token/ERC20/IERC20.sol";
+import { IZenBullStrategy } from "./interface/IZenBullStrategy.sol";
+import { IWETH9 } from "squeeth-monorepo/interfaces/IWETH9.sol";
+import { IEulerEToken } from "./interface/IEulerEToken.sol";
+import { IEulerDToken } from "./interface/IEulerDToken.sol";
+// contract
+import { ERC20 } from "openzeppelin/token/ERC20/ERC20.sol";
+import { UniFlash } from "./UniFlash.sol";
+import { UniOracle } from "./UniOracle.sol";
+// lib
+import { StrategyMath } from "squeeth-monorepo/strategy/base/StrategyMath.sol";
+import { Address } from "openzeppelin/utils/Address.sol";
+
+contract EmergencyWithdraw is ERC20, UniFlash {
+    using StrategyMath for uint256;
+    using Address for address payable;
+
+    uint256 internal constant ONE = 1e18;
+    /// @dev WETH decimals - USDC decimals
+    uint256 internal constant WETH_DECIMALS_DIFF = 1e12;
+    uint256 public constant MAX_WETH_PER_DEBT_REPAY = 100 ether;
+    uint256 public constant LIMIT_PRICE_TOLERANCE = 1e16; // 1%
+
+    /// @dev crab strategy address
+    address internal immutable crab;
+    /// @dev ZenBull strategy address
+    address internal immutable zenBull;
+    /// @dev WETH address
+    address internal immutable weth;
+    /// @dev USDC address
+    address internal immutable usdc;
+    /// @dev wPowerPerp address
+    address internal immutable wPowerPerp;
+    /// @dev eth/usdc uniswap pool address
+    address internal immutable ethUSDCPool;
+    address internal immutable eToken;
+    address internal immutable dToken;
+
+    /// @dev total amount of redeemed ZenBull through emergencyWithdrawEthFromCrab
+    uint256 public redeemedZenBullAmountForCrabWithdrawal;
+    /// @dev ZenBull total supply amount at the time of this contract deployment, used for Euler withdrawal calc
+    uint256 public redeemedRecoveryAmountForEulerWithdrawal;
+    /// @dev if true, enable ETH withdrawal from this contract
```

Comment:
maybe more clearer to say something like enable ETH withdrawal after 100% of Euler debt and collateral has been removed